openupgrade_scripts/scripts/sale/16.0.1.2/pre-migration.py

```diff
@@ -0,0 +1,46 @@
+from openupgradelib import openupgrade
+
+_models_renames = [
+    (
+        "sale.payment.acquirer.onboarding.wizard",
+        "sale.payment.provider.onboarding.wizard",
+    ),
+]
+
+_renames_xmlids = [
+    (
+        "sale.sale_payment_provider_onboarding_wizard_rule",
+        "sale.sale_payment_acquirer_onboarding_wizard_rule",
+    ),
+    (
+        "sale_management.menu_product_attribute_action",
+        "sale.menu_product_attribute_action",
+    ),
+]
+
+_moved_fields = [
+    "product_type",
+]
+
+
+def _noupdate_switch(env):
+    openupgrade.set_xml_ids_noupdate_value(
+        env, "sale", ["model_sale_order_action_share"], False
+    )
+
+
+def _remove_table_constraints(env):
+    openupgrade.delete_sql_constraint_safely(
+        env, "sale", "sale_order", "date_order_conditional_required"
+    )
+
+
+@openupgrade.migrate()
+def migrate(env, version):
+    openupgrade.update_module_moved_fields(
+        env.cr, "sale.order.line", _moved_fields, "sale_stock", "sale"
+    )
+    openupgrade.rename_models(env.cr, _models_renames)
+    openupgrade.rename_xmlids(env.cr, _renames_xmlids)
```

Comment:
an xmlid should also be created for the `ir.config_parameter` with key `"sale.default_invoice_email_template"`, because it exists in 15.0 but without xmlid, and an xmlid for it has been added in 16.0, which causes the migration to fail with this error:
```
bad query: INSERT INTO "ir_config_parameter" ("create_date", "create_uid", "key", "value", "write_date", "write_uid") VALUES ('2023-07-06 13:58:38.027885', 1, 'sale.default_invoice_email_template', '7', '2023-07-06 13:58:38.027885', 1) RETURNING "id"
ERROR: duplicate key value violates unique constraint "ir_config_parameter_key_uniq"
DETAIL: Key (key)=(sale.default_invoice_email_template) already exists.
```

here’s how to add the xmlid. maybe there is a cleaner way?
```python
    openupgrade.add_xmlid(
        env.cr,
        "sale",
        "default_invoice_email_template",
        "ir.config_parameter",
        env["ir.config_parameter"]
        .search([("key", "=", "sale.default_invoice_email_template")], limit=1)
        .id,
        noupdate=True,
    )
```

Reply:
@huguesdk Thanks for your help. I didn't realize this mistake

Reply:
with a database initialized with only `sale_stock`, this `ir.config_parameter` does not exist. sorry, my fix does not work in that case. instead, a test is needed, and it is better to move it to a function:

```python
def _create_xml_ids(env):
    config_parameter = env["ir.config_parameter"].search(
        [("key", "=", "sale.default_invoice_email_template")]
    )
    if config_parameter:
        openupgrade.add_xmlid(
            env.cr,
            "sale",
            "default_invoice_email_template",
            "ir.config_parameter",
            config_parameter.id,
            noupdate=True,
        )
```